lib/metric-config-parser/metric_config_parser/metric.py

```diff
@@ -239,6 +242,7 @@ def resolve(
                     owner=[self.owner] if isinstance(self.owner, str) else self.owner,
                     deprecated=self.deprecated,
                     level=self.level,
+                    aggregation_units=self.aggregation_units or [AnalysisUnit.CLIENT],
```

Comment:
We've got a few of these `self.parameter or [AnalysisUnit.CLIENT]` sprinkled around. What do you think about using attr's validators and converters instead (like [here](https://github.com/mozilla/mozanalysis/pull/273/files#diff-d3f3c2d9005e084f57601bb9527d9af10b834c513b2faf2dc5e926273b7b01e1R97-R102)). This would allow `Metric.__init__` to take a `aggregation_units=None` and convert that internally to `self.aggregation_units=[AnalysisUnit.CLIENT]`. It would move the default behavior logic internal to the class, instead of needing to be handled by callers.

Reply:
I actually started doing this on the DataSource side where we do use validators for some fields. But then I ended up deleting it because defining the type accomplished the same thing and there wasn't any additional validation needed. 

What's the `converter` do that `default` doesn't?

Also, if we do want to go down this road, I'd suggest we consider it separately and aim to cover more than just the new `aggregation_units` field.

Reply:
+1 to "let's do it later, not necessary for this PR"

As I understand it, `default` defines the default argument value while `converter` applies a function to the argument. Converters allow you to differentiate between the types of the parameters and the types of the attributes

So
```python
class MyClass:
  x = attr.field(default = None, type = str | None)
```
will define a class with a constructor like: 
```python
MyClass.__init__(self, x: str | None = None):
  self.x = x # type will be str | None
```

If you add a converter: 
```python
class MyClass: 
  x = attr.field(default = None, type = str, converter=attr.converter.default_if_none("default"))
```
you'll get a class with a constructor like: 
```python
MyClass.__init__(self, x: str | None = None):
  self.x = x if x is not None else "default" # type will be str, no longer optional
``` 

I think it comes down to a question of 1) what parameters do we want callers to be able to pass to the constructor and 2) what type do we want the class attributes to be.

Reply:
Got it, thanks for the examples.

I'm in favor of a bit more rigorously defined set of schemas for the various metric-hub entities. I was originally picturing pydantic schemas in nimbus-schemas package, but given that these aren't Nimbus specific that's not quite the right place for them. This might be a better approach (and a better location).

Reply:
I'm not that familiar with Pydantic. Do you think it's worth migrating to that as the solution?

Reply:
https://github.com/mozilla/metric-hub/issues/554

Reply:
Hm... I like pydantic, but my first thought is that it's not worth migrating. It was a bit of a pain incorporating pydantic stuff into jetstream for nimbus-schemas though partly due to the mix of attrs and pydantic, so having everything consistent would be nice. Seems like an ADR would be appropriate, though I'm not exactly itching to write it at the moment.

Reply:
Agree. Let's plan to circle back to this later, shouldn't block this PR.